src/storage/src/wal.rs

```diff
@@ -0,0 +1,413 @@
+
+        // Let's close the file descriptor from the old file and mark it as final.
+        old_file.sync_all().with_context(|| {
+            format!(
+                "failed to sync state for finished log segment: {}",
+                old_file_path.display()
+            )
+        })?;
+        fs::rename(&old_file_path, &old_file_rename).with_context(|| {
+            format!(
+                "failed to rename finished log segment from: {} to: {}",
+                old_file_path.display(),
+                old_file_rename.display()
+            )
+        })?;
+
+        // Update our own local state
+        self.current_sequence_number += 1;
+        self.current_bytes_written = 0;
+        Ok(())
+    }
+
+    /// Continue writing to the write-ahead log after a restart.
+    fn resume(id: GlobalId, base_path: PathBuf) -> Result<WriteAheadLog, anyhow::Error> {
+        let mut unfinished_file: Option<(PathBuf, usize)> = None;
+        // List out all of the files in the write-ahead log directory. There should
+        // be exactly one unfinished file, and potentially more than one finished
+        // file. If that's not the case, we need to error out.
+        let entries = std::fs::read_dir(&base_path)
+            .with_context(|| format!("failed to read wal directory {}", base_path.display()))?;
+        for entry in entries {
+            if let Ok(file) = entry {
+                let path = file.path();
+                let file_name = path.file_name();
+                if file_name.is_none() {
+                    continue;
+                }
+
+                let file_name = file_name.unwrap().to_str();
+
+                if file_name.is_none() {
+                    continue;
+                }
+
+                let file_name = file_name.unwrap();
+
+                if !file_name.starts_with("log-") {
+                    bail!(
+                        "found unexpected file when trying to resume wal for relation {}: {}",
+                        id,
+                        file_name,
+                    )
+                }
+
+                if !file_name.ends_with("final") {
+                    if unfinished_file.is_some() {
+                        bail!("found more than one in-progress log segments when trying to resume wal for relation {}: {} and {}",
+                            id,
+                            file_name,
+                            unfinished_file.unwrap().0.display())
+                    }
+                    let parts: Vec<_> = file_name.split('-').collect();
+                    let sequence_number = parts[1].parse::<usize>()?;
+                    unfinished_file = Some((path.to_path_buf(), sequence_number));
+                }
+            }
+        }
+
+        if let Some((unfinished_file, sequence_number)) = unfinished_file {
+            // Lets start writing to the previously created file.
+            let file = fs::OpenOptions::new()
```

Comment:
Should we use `create_log_segment` (probably renamed to `open_log_segment`) here to make sure we're always opening the files in an identical way?